src/lib/operators/table_scan.cpp

```diff
@@ -0,0 +1,206 @@
+    using Type = typename decltype(type)::type;
+
+    for (auto chunk_id = ChunkID{0}; chunk_id < chunk_count; ++chunk_id) {
+      auto position_list = std::make_shared<PosList>();
+      const auto chunk = input_table->get_chunk(chunk_id);
+      // if chunk is empty, skip this chunk
+      if (!chunk->size()) {
+        continue;
+      }
+      const auto segment = chunk->get_segment(_column_id);
+
+      const auto value_segment = std::dynamic_pointer_cast<ValueSegment<Type>>(segment);
+      const auto dictionary_segment = std::dynamic_pointer_cast<DictionarySegment<Type>>(segment);
+      const auto reference_segment = std::dynamic_pointer_cast<ReferenceSegment>(segment);
+
+      if (value_segment) {
+        position_list = _tablescan_value_segment<Type>(value_segment, chunk_id);
+      } else if (dictionary_segment) {
+        position_list = _tablescan_dict_segment<Type>(dictionary_segment, chunk_id);
+      } else if (reference_segment) {
+        position_list = _tablescan_reference_segment<Type>(reference_segment, chunk_id);
+      }
+
+      if (!position_list->empty()) {
+        output_reference_segments.push_back(std::make_shared<ReferenceSegment>(input_table, _column_id, position_list));
+      }
+    }
+  });
+
+  return std::make_shared<Table>(*input_table, output_reference_segments);
+}
+
+template <typename T>
+std::shared_ptr<PosList> TableScan::_tablescan_dict_segment(std::shared_ptr<DictionarySegment<T>> segment,
+                                                            ChunkID chunk_id) {
+  auto position_list = std::make_shared<PosList>();
+  const auto search_val = type_cast<T>(_search_value);
+
+  // Not comparing the individual types anymore, only indexes now.
+  auto scan_op = _create_scan_operation<ValueID>();
+
+  const auto lower_bound = segment->lower_bound(search_val);
+  const auto upper_bound = segment->upper_bound(search_val);
+
+  switch (_scan_type) {
+    /**
+     * Upper Bound is exclusive ==> [minElement, upperBound)
+     * Lower Bound is inclusive ==> [lowerBound, maxElement]
+     *
+     * Therefore, if they match, there is no matching element with the search value.
+     */
+    case ScanType::OpEquals:
+      if (lower_bound == upper_bound)
+        scan_op = [](auto l, auto r) { return false; };
+      break;
+    case ScanType::OpNotEquals:
+      if (lower_bound == upper_bound)
+        scan_op = [](auto l, auto r) { return true; };
+      break;
+    // Rest stays as it is.
+    case ScanType::OpLessThan:
+      break;
+    case ScanType::OpLessThanEquals:
+      break;
+    case ScanType::OpGreaterThan:
+      break;
+    case ScanType::OpGreaterThanEquals:
+      break;
+  }
+
+  const auto attr_vector = segment->attribute_vector();
+  for (auto value_ind = ChunkOffset{0}; value_ind < segment->size(); ++value_ind) {
+    if (scan_op(attr_vector->get(value_ind), lower_bound)) {
+      position_list->push_back(RowID{chunk_id, value_ind});
+    }
+  }
```

Comment:
```suggestion
  for (auto index = ChunkOffset{0}; index < segment->size(); ++index) {
    const auto value_id = attr_vector->get(index); 
    if (scan_op(value_id, lower_bound)) {
      position_list->push_back(RowID{chunk_id, index});
    }
  }
```

maybe this is easier to understand than "value_ind" being not the value_id but the index...

Reply:
agree